File: api/src/main/java/dev/akif/crud/simplest/SimplestService.java
package dev.akif.crud.simplest;

import dev.akif.crud.CRUDService;

import java.io.Serializable;
import java.time.Clock;

/**
 * The simplest variant of {@link CRUDService} where models are entities directly
 * <p>
 * This is meant to be extended from a {@link org.springframework.stereotype.Service} class.
 *
 * @param <I>      Id type of the data
 * @param <E>      Entity type of the data which is a {@link SimplestEntity}
 * @param <Mapper> Mapper type of the data which is a {@link SimplestMapper}
 */
public abstract class SimplestService<
        I extends Serializable,
        E extends SimplestEntity<I, E>,
        Mapper extends SimplestMapper<I, E>> extends CRUDService<I, E, E, E, E, Mapper> {
    /**
     * Constructor to provide type name and dependencies to this service
     *
     * @param type       Type name of the data this service manages
     * @param clock      {@link Clock} dependency of this service
     * @param repository Repository dependency of this service
     * @param mapper     Mapper dependency of this service
     */
    protected SimplestService(final String type, final Clock clock, final SimplestRepository<I, E> repository, final Mapper mapper) {
        super(type, clock, repository, mapper);
    }
}

File: api/src/main/kotlin/dev/akif/crud/simplest/SimplestService.kt
package dev.akif.crud.simplest

import dev.akif.crud.CRUDRepository
import dev.akif.crud.CRUDService
import dev.akif.crud.common.InstantProvider
import java.io.Serializable

/**
 * The simplest variant of [CRUDService] where models are entities directly
 *
 * This is meant to be extended from a **@Service** class.
 *
 * @param I      Id type of the data
 * @param E      Entity type of the data which is a [SimplestEntity]
 * @param R      Repository type of the data which is a [SimplestRepository]
 * @param Mapper Mapper type of the data which is a [SimplestMapper]
 */
abstract class SimplestService<
        I : Serializable,
        E : SimplestEntity<I, E>,
        out R : SimplestRepository<I, E>,
        out Mapper : SimplestMapper<I, E>>(
    override val typeName: String,
    override val instantProvider: InstantProvider,
    crudRepository: CRUDRepository<I, E>,
    override val mapper: Mapper
) : CRUDService<I, E, E, E, E, R, Mapper>(typeName, instantProvider, crudRepository, mapper)
